src/line.rs

```diff
@@ -59,30 +57,42 @@ impl Line {
         }
     }
 
-    fn insert_cursors(&self, text: &mut String, real_cursor: Option<&Cursor>) {
-        if let Some(ref cursors) = self.cursors {
-            for idx in cursors.iter().rev() {
-                let idx = *idx;
-                // If this cursor is the real cursor we don't want to draw it.
-                // We skip it, and the cursor will be set here later.
-                if let Some(real_cursor) = real_cursor {
-                    if real_cursor.column == idx {
-                        continue;
-                    }
-                }
+    fn add_styles(&self, text: &mut String) -> Result<()> {
+        if self.styles.is_empty() {
+            return Ok(());
+        }
+        if self.styles.len() % 3 != 0 {
+            // FIXME: find a more descriptive error for this. Hopefully, this should not happen
+            // though.
+            error!("Invalid style");
+            bail!(ErrorKind::DisplayError);
+        }
+        let mut style_idx = 0;
+        loop {
+            let start = self.styles[style_idx] as usize;
+            let mut end = start + self.styles[style_idx + 1] as usize;
 
-                // Make sure the cursor is within the bounds of the string by adding some padding.
-                // Note that if this happens, it can only happen for the right-most cursor.
-                if idx as usize + 1 > text.len() {
-                    for _ in 0..idx as usize + 1 - text.len() {
-                        text.push(' ');
-                    }
-                }
+            if end > text.len() {
+                // XXX: when dragging to a line below, xi-core sends a style that ends 1 index
```

Comment:
After thinking more about this, I really think this is normal behaviour, and we just need to handle it.
I'll remove this comment.